go/vt/worker/sharding_key.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2015, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package worker
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/youtube/vitess/go/sqltypes"
+
+	"github.com/youtube/vitess/go/vt/key"
+	"github.com/youtube/vitess/go/vt/mysqlctl/tmutils"
+	tabletmanagerdatapb "github.com/youtube/vitess/go/vt/proto/tabletmanagerdata"
+	topodatapb "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/topo"
+)
+
+// This file defines the interface and implementations of sharding key resolvers.
+
+// shardingKeyResolver defines the interface that needs to be satisifed to get a sharding
+// key from a database row.
+type shardingKeyResolver interface {
+	// shardingKey takes a table row, and returns the sharding key as bytes.
+	// It will return an error if no sharding key can be found.
+	shardingKey(tableName string, row []sqltypes.Value) ([]byte, error)
```

Comment:
The resolver should already know the table, we should have one constructed for a given table and just use that. That way it can be optimized to do all pre-computation before hands, and no lookup.

Maybe add an Init(TopoServer, keyspace, tableName) method in the interface (and guarantee it can only be called once), or add a factory function, either way, but we don't want any lookup in a map for every row...